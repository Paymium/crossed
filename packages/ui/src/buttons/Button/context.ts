/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { createContext } from 'react';
import { ButtonProps } from './types';

export const buttonContext = createContext<
  Pick<ButtonProps, 'variant' | 'size' | 'disabled' | 'error'> & {
    state?: {
      active?: boolean;
      hover?: boolean;
    };
    textId: string;
    // eslint-disable-next-line no-unused-vars
    setTextId: (p: string) => void;
  }
>({} as any);
