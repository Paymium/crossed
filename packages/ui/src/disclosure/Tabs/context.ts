/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { createScope } from '@crossed/core';
import { ScrollView } from 'react-native-gesture-handler';
import { SharedValue } from 'react-native-reanimated';

export type TabsContext = {
  value: string | number;
  setValue: (_value: string | number) => void;
  id: string;
  variant?: 'underline' | 'rounded';
  size?: 'sm' | 'md' | 'lg';
  listTabRef: React.MutableRefObject<ScrollView>;
  indicator: { left: SharedValue<number>; width: SharedValue<number> };
  scroll: SharedValue<number>;
  shouldShow: boolean;
  setShow: React.Dispatch<React.SetStateAction<boolean>>;
  widthLayout: SharedValue<number>;
};

export type TriggerContext = {
  disabled?: boolean;
  hover?: boolean;
  selected?: boolean;
};

export const createContext = () => {
  return {
    tabsContext: createScope<TabsContext>({
      variant: 'rounded',
    } as TabsContext),
    triggerContext: createScope<TriggerContext>({}),
  };
};
