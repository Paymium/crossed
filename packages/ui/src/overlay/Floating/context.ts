/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { createScope } from '@crossed/core';

export type FloatingContext = {
  open: boolean;
  onClose: () => void;
  onOpen: () => void;
  visibilityHidden?: boolean;
};

export const [FloatingProvider, useFloatingContext] =
  createScope<FloatingContext>({} as FloatingContext);
