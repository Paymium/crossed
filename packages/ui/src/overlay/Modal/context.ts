/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { createContext } from 'react';

export type VariantSize = { size: 'sm' | 'md' | 'lg' };

export const localContext = createContext<VariantSize>({ size: 'md' });
