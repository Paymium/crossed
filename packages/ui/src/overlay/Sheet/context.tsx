/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { createContext, ForwardedRef, RefObject, useContext } from 'react';
import type { SharedValue } from 'react-native-reanimated';
import { ActionSheetRef } from '@crossed/sheet';

export type SheetContext = {
  open: boolean;
  setOpen: (_value: boolean) => void;
  dismissOnOverlayPress?: boolean;
  hideHandle?: boolean;
  isMove: SharedValue<boolean>;
  height: SharedValue<number>;
  onClose: () => void;
  snapInitialHeight: SharedValue<number>;
  offset: number;
  ref: ForwardedRef<ActionSheetRef>;
  actionSheetRef: RefObject<ActionSheetRef>;
};

export const sheetContext = createContext<SheetContext>({} as SheetContext);

export const useSheetContext = () => useContext(sheetContext);
