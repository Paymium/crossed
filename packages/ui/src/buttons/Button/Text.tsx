/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { forwardRef, useContext, useEffect } from 'react';
import { Text, type TextProps } from '../../typography/Text';
import { buttonContext } from './context';
import { composeStyles } from '@crossed/styled';
import {
  buttonErrorStyles,
  buttonPrimaryStyles,
  buttonSecondaryStyles,
  buttonSuccessStyles,
  buttonTertiaryStyles,
  textStyles,
} from './styles';
import { Text as TextNative } from 'react-native';

type ButtonTextProps = TextProps;

const ButtonText = forwardRef<TextNative, ButtonTextProps>((props, ref) => {
  const { variant, state, disabled, setTextId, textId } =
    useContext(buttonContext);

  const { hover, active } = state || {};

  useEffect(() => {
    if (props.id && textId !== props.id) {
      setTextId(props.id);
    }
  }, [props.id, textId, setTextId]);

  return (
    <Text
      fontWeight={'lg'}
      {...props}
      id={textId}
      style={composeStyles(
        textStyles.default,
        variant === 'primary' && buttonPrimaryStyles.text,
        variant === 'primary' && hover && buttonPrimaryStyles.textHover,
        variant === 'primary' && active && buttonPrimaryStyles.textActive,
        variant === 'secondary' && buttonSecondaryStyles.text,
        variant === 'secondary' && hover && buttonSecondaryStyles.textHover,
        variant === 'secondary' && active && buttonSecondaryStyles.text,
        variant === 'tertiary' && buttonTertiaryStyles.text,
        variant === 'tertiary' && hover && buttonTertiaryStyles.textHover,
        variant === 'tertiary' && active && buttonTertiaryStyles.textActive,
        variant === 'error' && buttonErrorStyles.text,
        variant === 'success' && buttonSuccessStyles.text,
        disabled && textStyles.disabled,
        props.style
      )}
      ref={ref}
    />
  );
});
ButtonText.displayName = 'Button.Text';

export { ButtonText, type ButtonTextProps };
