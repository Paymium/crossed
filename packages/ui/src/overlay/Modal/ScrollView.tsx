/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { memo, useContext } from 'react';
import { ScrollView as RNScrollView } from 'react-native';
import { localContext } from './context';
import { Sheet } from '../Sheet';

export const ModalScrollView = memo((props) => {
  const { showSheet } = useContext(localContext);

  return showSheet ? (
    <Sheet.ScrollView {...props} />
  ) : (
    <RNScrollView {...props} />
  );
});
