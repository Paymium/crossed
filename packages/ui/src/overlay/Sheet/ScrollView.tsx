/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import {
  Children,
  forwardRef,
  PropsWithChildren,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import {
  type CrossedMethods,
  composeStyles,
  createStyles,
  inlineStyle,
} from '@crossed/styled';
import {
  type AnimatedScrollViewProps,
  runOnJS,
  useAnimatedProps,
  useAnimatedReaction,
  useAnimatedScrollHandler,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';
import { useSheetContext } from './context';
import { Handle } from './Handle';
import { Gesture, GestureDetector } from 'react-native-gesture-handler';
import { composeRefs } from '@crossed/core';
import { sheetStyles } from '../styles';
import { useFloatingContext } from '../Floating/context';
import { useMaxHeight } from './useMaxHeight';
import { ScrollView as SV } from '../../other/ScrollView';
import { SheetTitle } from './Title';
import { SheetFooter } from './Footer';
import { FlatList } from 'react-native';
import { YBox } from '../../layout/YBox';
import { useGesturePan } from './useGesturePan';
import { useGesture } from './useGesture';

const styles = createStyles(({ space, colors }) => ({
  container: {
    base: {
      paddingTop: space.xxs,
      gap: space.md,
    },
    variants: {},
  },
  containerPaddingBottom: { base: { paddingBottom: space.md } },
  containerPadded: {
    base: {
      paddingHorizontal: space.md,
    },
    variants: {},
  },
  fixed: {
    web: { base: { position: 'fixed' as any, boxSizing: 'content-box' } },
  },
  maxHeight: { base: { maxHeight: '100%' } },
  titleStyles: {
    web: {
      base: {
        background: `linear-gradient(${colors.background.secondary} 90%, transparent)`,
      },
    },
  },
  footer: {
    base: {
      paddingHorizontal: space.md,
      background: `linear-gradient(transparent 10%, ${colors.background.secondary})`,
    },
  },
}));

export type SheetScrollViewProps = Omit<AnimatedScrollViewProps, 'style'> & {
  style?: CrossedMethods<any, any>;
  padded?: boolean;
};

export const ScrollView = forwardRef<
  FlatList<any>,
  PropsWithChildren<Omit<SheetScrollViewProps, 'children'>>
>(({ children, style: styleProps, padded = true }, ref) => {
  const {
    hideHandle,
    snapInitialHeight,
    full,
    stickyFooter,
    stickyHeader,
    detach,
  } = useSheetContext();
  const { open, onClose } = useFloatingContext();

  const scrollRef = useRef<FlatList<any>>(null);

  const isMove = useSharedValue(false);
  const height = useSharedValue(0);
  const heightLayout = useSharedValue(0);
  const scroll = useSharedValue(0);
  const initialHeight = useSharedValue(0);
  const scrollViewEnable = useSharedValue(false);
  const [scrollEnabled, setScrollEnabled] = useState(false);

  const maxHeight = useMaxHeight();

  const native = Gesture.Native();
  const { gesturePan, styleAnimated } = useGesture({
    isMove,
    height,
    scroll,
    initialHeight,
    setScrollEnabled,
    scrollEnabled,
    snapInitialHeight,
    heightLayout,
  });

  useAnimatedReaction(
    () => {
      return snapInitialHeight.value
        ? snapInitialHeight.value
        : heightLayout.value >= maxHeight
          ? maxHeight
          : heightLayout.value;
    },
    (currentValue, previousValue) => {
      if (currentValue !== previousValue) {
        if (open && Math.floor(height.value) !== Math.floor(currentValue)) {
          runOnJS(setScrollEnabled)(currentValue === maxHeight);
          height.value = currentValue;
        }
      }
    },
    [heightLayout, open, maxHeight]
  );

  const onContentSizeChange = useCallback(
    (_w: number, h: number) => {
      heightLayout.value = full ? maxHeight : Math.floor(h);
    },
    [full, maxHeight, heightLayout]
  );

  const onScroll = useAnimatedScrollHandler((e) => {
    scroll.value = e.contentOffset.y;
  });

  const title = useMemo(() => {
    if (!children || typeof children === 'number') return null;
    return Children.toArray(children).find(
      (e) => typeof e === 'object' && 'type' in e && e.type === SheetTitle
    );
  }, [children]);
  const footer = useMemo(() => {
    if (!children) return null;
    return Children.toArray(children).find(
      (e) => typeof e === 'object' && 'type' in e && e.type === SheetFooter
    );
  }, [children]);
  const body = useMemo(() => {
    if (!children) return null;
    return Children.toArray(children).filter(
      (e) =>
        typeof e === 'object' &&
        'type' in e &&
        e.type !== SheetFooter &&
        e.type !== SheetTitle
    );
  }, [children]);

  return (
    <GestureDetector gesture={Gesture.Simultaneous(gesturePan, native)}>
      <SV
        ref={composeRefs(ref, scrollRef)}
        scrollEventThrottle={16}
        scrollEnabled={scrollEnabled}
        onScroll={onScroll}
        onContentSizeChange={onContentSizeChange}
        contentContainerStyle={
          composeStyles(
            padded && styles.container,
            !stickyFooter && styles.containerPaddingBottom,
            padded && styles.containerPadded
          ).style().style
        }
        style={{ zIndex: 1 }}
        containerProps={{
          style: [
            composeStyles(
              sheetStyles.content,
              sheetStyles.padding,
              detach &&
                inlineStyle(({ space }) => ({
                  base: {
                    right: space.xs,
                    left: space.xs,
                    bottom: space.xs,
                    borderBottomLeftRadius: 24,
                    borderBottomRightRadius: 24,
                  },
                })),
              styleProps || false
            ).style().style,
            { zIndex: 1 },
            styleAnimated,
          ],
        }}
        stickyHeader={stickyHeader}
        stickyFooter={stickyFooter}
      >
        <SV.Title>
          <YBox style={styles.titleStyles}>
            {!hideHandle && <Handle />}
            {title}
          </YBox>
        </SV.Title>
        <SV.Body>{body}</SV.Body>
        {Boolean(footer) && <SV.Footer>{footer}</SV.Footer>}
      </SV>
    </GestureDetector>
  );
});

ScrollView.displayName = 'SheetScrollView';
