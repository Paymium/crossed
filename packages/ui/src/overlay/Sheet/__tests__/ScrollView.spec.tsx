/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import React from 'react';
import { ScrollView } from '../ScrollView';
import { render, screen } from '@crossed/test';
import { sheetContext } from '../context';
import { expect, describe, jest, test } from '@jest/globals';
import { Text } from '../../../typography/Text';

describe('ScrollView', () => {
  test('not open', async () => {
    const snapInitialHeight = { value: 0 };
    const offset = 0;
    const open = false;
    const isMove = { value: 0 };
    const hideHandle = false;
    const height = { value: 0 };
    const onClose = jest.fn();
    render(
      <sheetContext.Provider
        value={
          {
            snapInitialHeight,
            offset,
            open,
            isMove,
            hideHandle,
            height,
            onClose,
          } as any
        }
      >
        <ScrollView testID="test">
          <Text>toto</Text>
        </ScrollView>
      </sheetContext.Provider>
    );
    expect(await screen.findByTestId('test')).toMatchSnapshot();
  });
  test('open', async () => {
    const snapInitialHeight = { value: 20 };
    const offset = 0;
    const open = true;
    const isMove = { value: 0 };
    const hideHandle = false;
    const height = { value: 300 };
    const onClose = jest.fn();
    render(
      <sheetContext.Provider
        value={
          {
            snapInitialHeight,
            offset,
            open,
            isMove,
            hideHandle,
            height,
            onClose,
          } as any
        }
      >
        <ScrollView testID="test">
          <Text>toto</Text>
        </ScrollView>
      </sheetContext.Provider>
    );
    expect(await screen.findByTestId('test')).toMatchSnapshot();
  });
});
