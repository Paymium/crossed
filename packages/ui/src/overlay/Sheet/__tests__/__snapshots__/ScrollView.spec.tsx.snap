// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollView not open 1`] = `
<div
  class="css-view-175oi2r r-WebkitOverflowScrolling-150rngu r-flexDirection-eqz5dr r-flexGrow-16y2uox r-flexShrink-1wbh5a2 r-overflowX-11yh6sk r-transform-agouwx"
  data-testid="test"
  style="position: fixed; bottom: 0px; right: 0px; left: 0px; z-index: 100000; border-top-left-radius: 24px; border-top-right-radius: 24px; max-height: 100%; box-sizing: content-box; height: 0px; overflow-y: auto; webkit-user-select: none; user-select: none;"
>
  <div
    class="css-view-175oi2r"
  >
    <div
      class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73 justify-content-[center] height-[40px]"
      tabindex="0"
    >
      <div
        class="css-view-175oi2r border-style-[solid] border-width-[0px] align-items-[stretch] background-color-[var(--colors-primary-20)] display-[flex] flex-basis-[auto] flex-direction-[column] flex-shrink-[0] list-style-[none] margin-[0px] min-height-[0px] min-width-[0px] padding-[0px] position-[relative] text-decoration-[none] z-index-[0] width-[54px] height-[5px] align-self-[center] border-radius-[14px] box-sizing-[border-box]"
      />
    </div>
    <div
      class="css-text-146c3p1 font-size-[16px] line-height-[20px] font-weight-[var(--font-font-weight-md)] color-[var(--colors-text-primary)] font-family-[var(--font-family)]"
      dir="auto"
    >
      toto
    </div>
  </div>
</div>
`;

exports[`ScrollView open 1`] = `
<div
  class="css-view-175oi2r r-WebkitOverflowScrolling-150rngu r-flexDirection-eqz5dr r-flexGrow-16y2uox r-flexShrink-1wbh5a2 r-overflowX-11yh6sk r-transform-agouwx"
  data-testid="test"
  style="position: fixed; bottom: 0px; right: 0px; left: 0px; z-index: 100000; border-top-left-radius: 24px; border-top-right-radius: 24px; max-height: 100%; box-sizing: content-box; height: 300px; overflow-y: hidden; webkit-user-select: none; user-select: none;"
>
  <div
    class="css-view-175oi2r"
  >
    <div
      class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73 justify-content-[center] height-[40px]"
      tabindex="0"
    >
      <div
        class="css-view-175oi2r border-style-[solid] border-width-[0px] align-items-[stretch] background-color-[var(--colors-primary-20)] display-[flex] flex-basis-[auto] flex-direction-[column] flex-shrink-[0] list-style-[none] margin-[0px] min-height-[0px] min-width-[0px] padding-[0px] position-[relative] text-decoration-[none] z-index-[0] width-[54px] height-[5px] align-self-[center] border-radius-[14px] box-sizing-[border-box]"
      />
    </div>
    <div
      class="css-text-146c3p1 font-size-[16px] line-height-[20px] font-weight-[var(--font-font-weight-md)] color-[var(--colors-text-primary)] font-family-[var(--font-family)]"
      dir="auto"
    >
      toto
    </div>
  </div>
</div>
`;
